src/NHibernate.Test/NHSpecificTest/NH3932/Model.cs

```diff
@@ -252,7 +275,7 @@ public override bool Equals(object obj)
 
 		public override int GetHashCode()
 		{
-			return 1; //for simlicity - care about equals impl
+			return Id;
```

Comment:
With this change, we now have following expression on the parent set, after the `ISession.Save`, yielding `false`:
```C#
((SetParent)storedParent).Children.Contains(System.Linq.Enumerable.First(((SetParent)storedParent).Children))
```
Which is wrong, of course. This happens because the child has been added in the set with `0` as hashcode but now yields `1` as hashcode, causing the set to no more knows it already contains it.

It does not trigger a bug in current test cases because the way `GenericSetType.AreCollectionElementsEqual` is currently coded, the test is done on the cloned set, which has stored the children with its latest hashcode. Inverting on which set the `SetEquals` call is done triggers the bug.

The usual way to avoid this trouble is to compute the hashcode only once per instance, storing it and then always yielding the first computed value. (A `Lazy<int>` behavior.) This does not fix all potential troubles since querying the child from a new session would result in obtaining a clone child, being equal but with a different hashcode. In the case of your test, this causes the merging of the "clean clone" (which only clone the parent, not the children) to fail due to this.

The safer in the case of your tests would be to reload the to be stored parent entity from a new session before storing it in the fixture. Or use the `assigned` id generator and assign child id before adding it to a set.